### PR TITLE
Refactor promises in throws method

### PR DESCRIPTION
**Pull request by SamVerschueren**
Regardless of issue https://github.com/sindresorhus/ava/issues/123, I think the promise code in `x.throws` might be written more cleaner.

### CODE REVIEW

**Comment by SamVerschueren** [comment]
This is odd, because of this minor change, 40 tests are failing.

**Comment by schnittstabil** [comment]
@SamVerschueren see my comment [here](https://github.com/sindresorhus/ava/pull/91) if you want to fix that.

File: lib/assert.js [hunk 1]
```diff
@@ -70,19 +70,11 @@ x.notSame = function (val, expected, msg) {
...
+			.then(function () {
+				x.throws(function () {}, err, msg);
 			})
-			.then(function (fnErr) {
-				if (!isFailed) {
-					x.throws(function () {}, err, msg);
-				}
-
+			.catch(function (fnErr) {
 				x.throws(function () {
```
Comment (schnittstabil) [inline lib/assert.js]: That will catch errors from `fn` **and** errors thrown at line 75…

File: lib/assert.js [hunk 2]
```diff
@@ -70,19 +70,11 @@ x.notSame = function (val, expected, msg) {
...
-				return err;
+			.then(function () {
+				x.throws(function () {}, err, msg);
 			})
-			.then(function (fnErr) {
-				if (!isFailed) {
-					x.throws(function () {}, err, msg);
-				}
-
+			.catch(function (fnErr) {
```
Comment (schnittstabil) [inline lib/assert.js]: That will catch errors from `fn` **and** errors thrown at line 75…

Comment (SamVerschueren) [inline lib/assert.js]: In what situation does line 75 throw an error? It seems you are correct though, if I change it to your suggestion, it works.

```js
			.then(function () {
				x.throws(function () {}, err, msg);
			}, function (fnErr) {
				x.throws(function () {
					throw fnErr;
				}, err, msg);
			});
```

Comment (schnittstabil) [inline lib/assert.js]: `x.throws(function () {}, err, msg);` throws an error, and is evaluated if `fn` isn't rejected.

**Comment by vdemedes** [comment]
Could you update it to:

```js
return fn
  .catch()
  .then()
```

instead of shortcut form?

**Comment by sindresorhus** [comment]
@vdemedes See: https://github.com/sindresorhus/ava/pull/124#discussion_r43855517

**Comment by vdemedes** [comment]
@sindresorhus this applies if we have `.then()` and `.catch()` in that order.

`.then(yes, no)` is a shortcut for `.catch(no).then(yes)`.

**Comment by schnittstabil** [comment]
@vdemedes It is not a shortcut:
```js
Promise.reject('err').catch(() => null).then(() => console.log('success'))
//=> outputs: success

Promise.resolve(() => console.log('success'), () => null)
//=> no output
```

**Comment by vdemedes** [comment]
@schnittstabil https://github.com/petkaantonov/bluebird/wiki/Promise-anti-patterns#the-thensuccess-fail-anti-pattern

**Comment by vdemedes** [comment]
My suggestion is to convert this:

```js
.then(function () {
			x.throws(function () {}, err, msg);
			}, function (fnErr) {
 				x.throws(function () {
 					throw fnErr;
 				}, err, msg);
 			});
```

to this:
```js
.catch(function (fnErr) {
	x.throws(function () {
		throw fnErr;
	}, err, msg);
})
.then(function () {
	x.throws(function () {}, err, msg);
})
```

**Comment by schnittstabil** [comment]
> …  and rationalize to yourself that at least the code is "less coupled" or something.
I don't do that :bowtie:

**Comment by schnittstabil** [comment]
@vdemedes if you really, really, really want to avoid `.then(…, …)` then you may prefer:

```js
	if (fn && fn.then) {
		return new Promise(function (resolve, reject) {
			fn.then(function () {
				try {
					x.throws(function () {}, err, msg);
				} catch (e) {
					reject(e);
				}
			}).catch(function (errFn) {
				try {
					x.throws(function () {
						throw errFn;
					}, err, msg);
					resolve();
				} catch (e) {
					reject(e);
				}
			});
		});
	}
```

**Comment by SamVerschueren** [comment]
@schnittstabil The reason I created this PR was to make the code cleaner, not make it more complex :). So your example is definitely a no-go.

@vdemedes Suggestion doesn't work

#### Example

```js
var Promise = require('bluebird');

Promise.reject(new Error('foo'))
    .catch(function (err) {
        console.log(err.message);
    })
    .then(function (success) {
        console.log(success);
    });
// => foo
// => undefined

Promise.resolve('bar')
    .catch(function (err) {
        console.log(err.message);
    })
    .then(function (success) {
        console.log(success);
    });
// => bar

Promise.reject(new Error('foo'))
    .then(function (success) {
        console.log(success);
    }, function (err) {
        console.log(err.message);
    });
// => foo

Promise.resolve('bar')
    .then(function (success) {
        console.log(success);
    }, function(err) {
        console.log(err.message);
    });
// => bar
```

It looks like the only option is the `then(success, err)` one. If you don't want to use that one, just keep it the way it is :).

**Comment by sindresorhus** [comment]
> It looks like the only option is the then(success, err) one. 

I'm fine with using this pattern. It's IMHO only an anti-pattern when you use it thinking it's a shortcut for `.catch()`.

**Comment by vdemedes** [comment]
Ok, let's merge this and we'll see what we can do later.

**Pull request closed without merging**